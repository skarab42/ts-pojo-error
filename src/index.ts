// ---------------------------------------------------------------------------
// Pojo Types
// ---------------------------------------------------------------------------
export type Unwrap<TType> = TType extends Record<string, unknown>
  ? { [TKey in keyof TType]: TType[TKey] }
  : TType;

export type PojoErrorPayload = { message: string };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type PojoErrorCallback = (...args: any) => PojoErrorPayload;

export type PojoErrorTypes = Record<string, PojoErrorCallback>;

export type PojoEnum<TErrorTypes> = { [TKey in keyof TErrorTypes]: TKey };

export type PojoFakeEnum<TErrorTypes> = {
  enumObject: PojoEnum<TErrorTypes>;
  enumKeys: [keyof TErrorTypes];
};

export type PojoErrorInstance<
  TErrorTypes extends PojoErrorTypes,
  TType extends keyof TErrorTypes,
> = PojoError<
  TType,
  Parameters<TErrorTypes[TType]>,
  ReturnType<TErrorTypes[TType]>
>;

export type NewPojoError<TErrorTypes extends PojoErrorTypes> = <
  TType extends keyof TErrorTypes,
>(
  type: TType,
  ...args: [...Parameters<TErrorTypes[TType]>]
) => PojoErrorInstance<TErrorTypes, TType>;

export type ThrowPojoError<TErrorTypes extends PojoErrorTypes> =
  NewPojoError<TErrorTypes>;

export type NewPojoErrorWithCause<TErrorTypes extends PojoErrorTypes> = <
  TType extends keyof TErrorTypes,
>(
  cause: Error,
  type: TType,
  ...args: [...Parameters<TErrorTypes[TType]>]
) => PojoErrorInstance<TErrorTypes, TType>;

export type ThrowPojoErrorWithCause<TErrorTypes extends PojoErrorTypes> =
  NewPojoErrorWithCause<TErrorTypes>;

export type PojoFactory<TErrorTypes extends PojoErrorTypes> = {
  type: Unwrap<PojoEnum<TErrorTypes>>;
  errors: TErrorTypes;
  new: NewPojoError<TErrorTypes>;
  newFrom: NewPojoErrorWithCause<TErrorTypes>;
  throw: ThrowPojoError<TErrorTypes>;
  throwFrom: NewPojoErrorWithCause<TErrorTypes>;
  is: <TType extends keyof TErrorTypes>(
    type: TType,
    error: unknown,
  ) => error is PojoErrorInstance<TErrorTypes, TType>;
  has: <TType extends keyof TErrorTypes>(
    error: unknown,
  ) => error is PojoErrorInstance<TErrorTypes, TType>;
};

export type PojoObject<
  TType,
  TArgs extends [unknown, ...unknown[]],
  TPojo extends PojoErrorPayload,
> = {
  type: TType;
  args: TArgs;
  data: TPojo;
  stack: string | undefined;
};

// ---------------------------------------------------------------------------
// PojoError Class
// ---------------------------------------------------------------------------
export class PojoError<
  TType,
  TArgs extends [unknown, ...unknown[]],
  TPojo extends PojoErrorPayload,
> extends Error {
  readonly type: TType;
  readonly args: TArgs;
  readonly data: TPojo;
  readonly cause?: Error | undefined;

  constructor(
    type: TType,
    args: TArgs,
    data: TPojo,
    options: { cause?: Error; constructorOpt?: Function },
  ) {
    super();

    this.type = type;
    this.args = args;
    this.data = data;
    this.cause = options.cause;
    this.message = data.message;

    if (typeof Error.captureStackTrace === "function") {
      Error.captureStackTrace(this, options.constructorOpt);
    }

    Object.setPrototypeOf(this, new.target.prototype);
  }

  toObject(): Unwrap<PojoObject<TType, TArgs, TPojo>> {
    return {
      type: this.type,
      args: [...this.args],
      data: { ...this.data },
      stack: this.stack,
    };
  }

  toJSON(): string {
    return JSON.stringify(this.toObject());
  }
}

// ---------------------------------------------------------------------------
// PojoError Factory
// ---------------------------------------------------------------------------
function fakeEnum<TErrorTypes extends PojoErrorTypes>(
  errors: TErrorTypes,
): PojoFakeEnum<TErrorTypes> {
  const enumKeys = Object.keys(errors) as [keyof TErrorTypes];
  const enumObject = Object.create(null) as PojoEnum<TErrorTypes>;

  enumKeys.forEach((key) => (enumObject[key] = key));

  return { enumObject, enumKeys };
}

export function factory<TErrorTypes extends PojoErrorTypes>(
  errors: TErrorTypes,
): Unwrap<PojoFactory<TErrorTypes>> {
  const { enumObject } = fakeEnum(errors);

  function newError<TType extends keyof TErrorTypes>(
    type: TType,
    ...args: [...Parameters<TErrorTypes[TType]>]
  ): PojoErrorInstance<TErrorTypes, TType> {
    const func = errors[type] as TErrorTypes[TType];
    const data = func(...args) as ReturnType<TErrorTypes[TType]>;

    return new PojoError(type, args, data, { constructorOpt: newError });
  }

  function newFromError<TType extends keyof TErrorTypes>(
    cause: Error,
    type: TType,
    ...args: [...Parameters<TErrorTypes[TType]>]
  ): PojoErrorInstance<TErrorTypes, TType> {
    const func = errors[type] as TErrorTypes[TType];
    const data = func(...args) as ReturnType<TErrorTypes[TType]>;

    return new PojoError(type, args, data, { cause, constructorOpt: newError });
  }

  function throwError<TType extends keyof TErrorTypes>(
    type: TType,
    ...args: [...Parameters<TErrorTypes[TType]>]
  ): PojoErrorInstance<TErrorTypes, TType> {
    throw newError(type, ...args);
  }

  function throwFromError<TType extends keyof TErrorTypes>(
    cause: Error,
    type: TType,
    ...args: [...Parameters<TErrorTypes[TType]>]
  ): PojoErrorInstance<TErrorTypes, TType> {
    throw newFromError(cause, type, ...args);
  }

  function isError<TType extends keyof TErrorTypes>(
    type: TType,
    error: unknown,
  ): error is PojoErrorInstance<TErrorTypes, TType> {
    return error instanceof PojoError && error.type === type;
  }

  function hasError<TType extends keyof TErrorTypes>(
    error: unknown,
  ): error is PojoErrorInstance<TErrorTypes, TType> {
    return (
      error instanceof PojoError &&
      typeof enumObject[error.type as TType] !== "undefined"
    );
  }

  return {
    type: enumObject as Unwrap<typeof enumObject>,
    new: newError,
    newFrom: newFromError,
    throw: throwError,
    throwFrom: throwFromError,
    is: isError,
    has: hasError,
    errors,
  };
}
