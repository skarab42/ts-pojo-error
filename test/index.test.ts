/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { expect, test } from "vitest";
import { factory, PojoError } from "../src";
import { errors, ErrorsEnum, errorsEnum, sessionErrors } from "./fixtures";

function assertType<TExpected>(input: TExpected): TExpected {
  return input;
}

test("factory", () => {
  const myErrors = factory(errors);

  assertType<ErrorsEnum>(myErrors.type);
  assertType<typeof errors>(myErrors.errors);

  expect(myErrors.type).toEqual(errorsEnum);
  expect(myErrors.errors).toBe(errors);
});

test("factory.new", () => {
  const myErrors = factory(errors);

  myErrors.new("FATAL", "Oupsy!");
  myErrors.new(myErrors.type.FATAL);

  // @ts-expect-error Expected 2-3 arguments, but got 1.
  myErrors.new(myErrors.type.PAGE_NOT_FOUND);
  // @ts-expect-error Expected 1-2 arguments, but got 3.
  myErrors.new(myErrors.type.FATAL, "Oupsy!", 42);

  // @ts-expect-error Argument of type '"PROUT"' is not assignable ...
  expect(() => myErrors.new("PROUT")).toThrow();
});

test("factory.throw", () => {
  const myErrors = factory(errors);
  expect(() => myErrors.throw("FATAL")).toThrow("Fatal error");
});

test("factory.has", () => {
  const myErrors = factory(errors);
  const mySessionErrors = factory(sessionErrors);

  try {
    throw myErrors.new("FATAL");
  } catch (error) {
    expect(myErrors.has(error)).toBe(true);
    expect(mySessionErrors.has(error)).toBe(false);

    if (mySessionErrors.has(error)) {
      assertType<"LOGIN" | "AUTH">(error.type);
    }
  }
});

test("instanceof PojoError", () => {
  const myErrors = factory(errors);

  try {
    throw myErrors.new("FATAL");
  } catch (error) {
    expect(error instanceof PojoError).toBe(true);
  }
});

test("error without parameter", () => {
  const myErrors = factory(errors);
  const myError = myErrors.new(myErrors.type.FATAL);

  assertType<"FATAL">(myError.type);
  assertType<Parameters<typeof errors.FATAL>>(myError.args);
  assertType<ReturnType<typeof errors.FATAL>>(myError.data);

  try {
    throw myError;
  } catch (error) {
    expect(error).instanceOf(PojoError);

    if (myErrors.is("FATAL", error)) {
      assertType<"FATAL">(error.type);
      assertType<Parameters<typeof errors.FATAL>>(error.args);
      assertType<ReturnType<typeof errors.FATAL>>(error.data);

      expect(error.type).toBe("FATAL");
      expect(error.args).toEqual([]);
      expect(error.data).toEqual({ message: "Fatal error" });
      expect(error.message).toBe("Fatal error");
    }
  }
});

test("error with one parameter", () => {
  const myErrors = factory(errors);
  const myError = myErrors.new(myErrors.type.FATAL, "My custom fatal error");

  try {
    throw myError;
  } catch (error) {
    expect(error).instanceOf(PojoError);

    if (myErrors.is("FATAL", error)) {
      expect(error.type).toBe("FATAL");
      expect(error.args).toEqual(["My custom fatal error"]);
      expect(error.data).toEqual({ message: "My custom fatal error" });
      expect(error.message).toBe("My custom fatal error");
    }
  }
});

test("error with two parameter (url, undefined)", () => {
  const myErrors = factory(errors);
  const myError = myErrors.new(
    myErrors.type.PAGE_NOT_FOUND,
    "http://www.skarab42.dev",
  );

  try {
    throw myError;
  } catch (error) {
    expect(error).instanceOf(PojoError);

    if (myErrors.is("PAGE_NOT_FOUND", error)) {
      expect(error.data).toMatchSnapshot();
      expect(error.type).toBe("PAGE_NOT_FOUND");
      expect(error.args).toEqual(["http://www.skarab42.dev"]);
      expect(error.message).toBe("Page Not Found: http://www.skarab42.dev");
    }
  }
});

test("error with two parameters (url, user)", () => {
  const user = { name: "skarab42", isAdmin: true };
  const myErrors = factory(errors);
  const myError = myErrors.new(
    myErrors.type.PAGE_NOT_FOUND,
    "http://www.skarab42.dev",
    user,
  );

  try {
    throw myError;
  } catch (error) {
    expect(error).instanceOf(PojoError);

    if (myErrors.is("PAGE_NOT_FOUND", error)) {
      expect(error.data).toMatchSnapshot();
      expect(error.type).toBe("PAGE_NOT_FOUND");
      expect(error.args).toEqual(["http://www.skarab42.dev", user]);
      expect(error.message).toBe("Page Not Found: http://www.skarab42.dev");
    }
  }
});

test("error to object/json", () => {
  const user = { name: "skarab42", isAdmin: true };
  const myErrors = factory(errors);
  const myError = myErrors.new(
    myErrors.type.PAGE_NOT_FOUND,
    "http://www.skarab42.dev",
    user,
  );

  const myErrorObject = myError.toObject();

  expect(myErrorObject).toMatchSnapshot({
    stack: expect.stringMatching(
      /^PojoError: Page Not Found: http:\/\/www\.skarab42\.dev/,
    ),
  });

  expect(JSON.parse(myError.toJSON())).toEqual(myErrorObject);
});

test("new error with cause", () => {
  const myErrors = factory(errors);

  try {
    throw myErrors.new("UNKNOWN");
  } catch (error) {
    if (myErrors.has(error)) {
      const newError = myErrors.newFrom(error, "FATAL");

      if (myErrors.is("FATAL", newError)) {
        expect(newError.cause).toBe(error);
      }
    }
  }
});

test("throw error with cause", () => {
  const myErrors = factory(errors);

  function action(): void {
    try {
      throw myErrors.new("UNKNOWN");
    } catch (error) {
      if (myErrors.has(error)) {
        myErrors.throwFrom(error, "FATAL");
      }
    }
  }

  try {
    action();
  } catch (error) {
    if (myErrors.is("FATAL", error)) {
      expect(myErrors.is("UNKNOWN", error.cause)).toBe(true);
    }
  }
});

test("error with cause to object", () => {
  const myErrors = factory(errors);

  try {
    throw new Error("Native Error!!!");
  } catch (error) {
    if (error instanceof Error) {
      const newError1 = myErrors.newFrom(error, "FATAL");
      const newError2 = myErrors.newFrom(newError1, "PAGE_NOT_FOUND", "www");
      const newError3 = myErrors.newFrom(newError2, "WARNING", "Danger !!!");

      const common = {
        type: expect.any(String),
        args: expect.any(Array),
        data: expect.any(Object),
      };

      expect(newError3.toObject()).toMatchSnapshot({
        ...common,
        stack: expect.stringMatching(/^PojoError: Danger !!!/),
        cause: {
          ...common,
          stack: expect.stringMatching(/^PojoError: Page Not Found/),
          cause: {
            ...common,
            stack: expect.stringMatching(/^PojoError: Fatal error/),
            cause: {
              ...common,
              stack: expect.stringMatching(/^Error: Native Error!!!/),
            },
          },
        },
      });
    }
  }
});
