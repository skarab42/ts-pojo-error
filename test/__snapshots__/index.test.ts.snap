// Vitest Snapshot v1

exports[`error to object/json 1`] = `
{
  "args": [
    "http://www.skarab42.dev",
    {
      "isAdmin": true,
      "name": "skarab42",
    },
  ],
  "cause": undefined,
  "data": {
    "message": "Page Not Found: http://www.skarab42.dev",
    "request": {
      "url": "http://www.skarab42.dev",
      "user": {
        "isAdmin": true,
        "name": "skarab42",
      },
    },
  },
  "stack": StringMatching /\\^PojoError: Page Not Found: http:\\\\/\\\\/www\\\\\\.skarab42\\\\\\.dev/,
  "type": "PAGE_NOT_FOUND",
}
`;

exports[`error with cause to object 1`] = `
{
  "args": Any<Array>,
  "cause": {
    "args": Any<Array>,
    "cause": {
      "args": Any<Array>,
      "cause": {
        "args": Any<Array>,
        "cause": undefined,
        "data": Any<Object>,
        "stack": StringMatching /\\^Error: Native Error!!!/,
        "type": Any<String>,
      },
      "data": Any<Object>,
      "stack": StringMatching /\\^PojoError: Fatal error/,
      "type": Any<String>,
    },
    "data": Any<Object>,
    "stack": StringMatching /\\^PojoError: Page Not Found/,
    "type": Any<String>,
  },
  "data": Any<Object>,
  "stack": StringMatching /\\^PojoError: Danger !!!/,
  "type": Any<String>,
}
`;

exports[`error with two parameter (url, undefined) 1`] = `
{
  "message": "Page Not Found: http://www.skarab42.dev",
  "request": {
    "url": "http://www.skarab42.dev",
    "user": undefined,
  },
}
`;

exports[`error with two parameters (url, user) 1`] = `
{
  "message": "Page Not Found: http://www.skarab42.dev",
  "request": {
    "url": "http://www.skarab42.dev",
    "user": {
      "isAdmin": true,
      "name": "skarab42",
    },
  },
}
`;
